### ISSUE TITLE
Question: Save method updating only the changed fields

### ISSUE THREAD

**Issue by andrefcgoncalves** [typeorm/typeorm#7255]
**Issue type:**

[x] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

When using the save method (using either the Active Record or Data Mapper pattern) all the fields that are different from the ones on the database at the moment of the save are updated, even if those fields were not changed on the entity.

I would expect the save method to only update the fields that were changed since the entity load from the database.

Let me give an example:
```typescript
// Create user
const user = new User;
user.email = 'user@email.com';
user.name = 'User';
await user.save();

// Get same user 2 times
const user1 = await User.findOneOrFail(1);
const user2 = await User.findOneOrFail(1);

user1.email = 'user1@email.com';

user2.name = 'User2';

await user1.save(); // Updates email to user1@email.com

await user2.save(); // updates name to User2 and email to user@email.com. It should only update the name
```

In the end, I would expect the data in the database to have `email='user1@email.com'`and `name = 'User2'` but that's not the case, as the second save reverts the update of the first one.

This is, of course, not a practical example, but just imagine that this code was divided into two endpoints (one to update the email and another to update the name) that were called simultaneously.

Looking at the source code, I noticed every time there's a save, a select query is done before the update and the result from that query is used to compare with the fields on the entity to be updated.
I believe this comparison should instead be done with the data that was retrieved from the database at the moment of the select query.

I was able to change the behaviour by creating an entity subscriber:
```typescript
@EventSubscriber()
export class GenericSubscriber implements EntitySubscriberInterface {
  afterLoad(entity: any) {
    // clone entity
    entity.databaseEntity = JSON.parse(JSON.stringify(entity));
  }
}
```
and by changing the code of the EntityPersistExecutor.ts:75 to the following:
```typescript
const subject = new Subject({
    metadata: this.connection.getMetadata(entityTarget),
    entity: entity,
    canBeInserted: this.mode === "save",
    canBeUpdated: this.mode === "save",
    mustBeRemoved: this.mode === "remove",
    canBeSoftRemoved: this.mode === "soft-remove",
    canBeRecovered: this.mode === "recover"
});
subject.databaseEntity = entity.databaseEntity;
subject.databaseEntityLoaded = true;
subjects.push(subject);
```

- Is the current behaviour actually the intended one or should this issue be treated as a bug?
- Is there any problem with the solution presented? And if not, it makes sense for this to be a typeorm configuration?

**Comment by nebkat** [comment]
> I would expect the save method to only update the fields that were changed since the entity load from the database.

> In the end, I would expect the data in the database to have email='user1@email.com'and name = 'User2' but that's not the case, as the second save reverts the update of the first one.

I would equally say that at any point you run `save()` with a full entity you would expect the database to match the state of that entity once completed. Don't forget that `save()` can also accept manually created objects, for which there would be no modification history, and behavior should be consistent no matter what kind of input is provided.

For that example `save()` wouldn't be the correct tool. Instead you should use `update()` which directly performs an `UPDATE` query (without comparing to existing rows), and you can then choose to only update the changed columns.

**Comment by andrefcgoncalves** [comment]
@nebkat Thank you for the reply.

I understand your point and it makes sense, of course. But I know that are some other ORMs that do it differently and that's why I was asking, just to make sure it was not a bug or to see if it made sense for this to be a configuration option.
I can't understand how the current implementation of this method can be used for applications whith multiple parallel requests and partial updates.

The thing with using the `update()` method is that I allways have to keep 2 separate objects: the entity object itself and another object with the fields that I want to update. I would probably had to pass both objects to every method that I would need to call.

Regarding my proposed solution, can you comment if it would make sense? I can make that change in my project but not sure if it would break something.

Thanks

**Comment by nebkat** [comment]
I think the correct way to do this would be using a custom base entity class, that keeps track of modifications to the object, perhaps using a `Proxy`. You would then have a function `entity.modifications(): Partial<Entity>` that returns an object with all the modified fields, which you could pass to `update()`.

Are these other ORMs also JavaScript based?

**Comment by andrefcgoncalves** [comment]
Yes, sequelize, for instance. I tested the same example in another project where I'm using it.

Thank you for the suggestion. I will try it that way.

**Comment by mathew-kurian** [comment]
@andrefcgoncalves Do you have any updates for this? An option to save only the fields you changed would be great! Useful in systems with highly contentious table entries.

**Comment by andrefcgoncalves** [comment]
@mathew-kurian Unfortunately, I do not have any updates. 

My original solution had problems with saving/updating relations and the solution suggested by @nebkat also wouldn't work for saving relations (at least the way we wanted).

**Comment by mathew-kurian** [comment]
@andrefcgoncalves ok got it. Can you please let me know what specifically in relations? (I am still a bit new to ORM)

**Comment by Eriasuitor** [comment]
I encountered the same situation and lots of occasional bugs have been incurred in the project I'm attempt to taking over.

**Comment by Eriasuitor** [comment]
I encountered the same situation and lots of occasional bugs have been incurred in the project I was taking over. These bugs are fatal in general.

**Comment by cool-firer** [comment]
It's 2025, encountered the same situation, didn't expect the save behavior to be different from Sequelize.
